### PR TITLE
fix: handle insufficient disk space gracefully

### LINKED ISSUE #575: Missing checks for insufficient disk space
Hi, 

I was doing some testing recently and noticed that we might be missing checks and handling for full disks (ENOSPC).

Currently, if a user's disk fills up mid-download, `file.WriteAt` (in `concurrent/worker.go`) and `io.CopyBuffer` (in `single/downloader.go`) fail. However, because disk-full isn't caught as a permanent error, it triggers a cascading retry storm:
1. In `concurrent`, the worker retries in-place up to `maxRetries` times, pulling the same chunk over the network again with mirror rotations/backoffs.
2. The chunk is then requeued and re-enters the worker pool rotation, repeating the cycle.
3. Finally, the scheduler requeues the entire task up to 10 times.

**Proposed Approach:**
Similar to how the engine handles `ErrPermanentHTTP` to skip pointless retries, we could:
1. Add an `ErrInsufficientDiskSpace` in `types/errors.go`.
2. Map write errors to this sentinel (e.g., checking `errno == syscall.ENOSPC` or `windows.ERROR_DISK_FULL`).
3. Check for this sentinel in workers and `scheduler.go` to instantly break out of the retry cycles.
4. *(Optional UX Enhancement)*: Building on this, we could perhaps issue a warning (or fail early) if a download is projected to occupy more than a configurable threshold of the available free space once the `Content-Length` is known.

Let me know your thoughts on this! I’d be happy to take a crack at a PR if it helps (though might need a few pointers on the TUI side). Alternatively, if you guys prefer to just take this and run with it yourselves, or if you have a different approach in mind, please feel free! Happy to support however I can.

### PR DESCRIPTION
**Pull request by superGekFordJ**
Hey @junaid2005p, thanks for taking a look at the issue at #575! 

Since you mentioned this freezing issue is a P0 priority, we took a crack at implementing a solution on our end for GoAria. Good news: we are now able to correctly intercept downloads that exceed available disk space at enqueue, and if the disk becomes full mid-download, they gracefully pause and correctly persist the interrupted progress. I've ported the solution back here so you guys can evaluate it.

While considering how to handle scenarios where the disk suddenly fills up mid-download, we realized the need for proper resume capability for tasks in the `error` state.
Currently, the concurrent downloader already correctly calls `saveStateSnapshot(emit=false)`, but there was no bridge to pass this snapshot up to the orchestrator. Because `events.go` lacked the `m.State != nil` handling for `EventError`, it never called `SaveStateWithOptions`. The result was that if a download failed (due to ENOSPC or any other error), the chunk bitmap was effectively dropped, preventing resume. This PR wires up that bridge.

### What this PR achieves:
*(The following summary was generated by AI to help organize the changes)*
1. **Retry Storm Suppression**: Mapped platform write errors to a new `types.ErrInsufficientDiskSpace` sentinel. Workers now early-return immediately, and the scheduler excludes it from `MaxTaskRetries`. This stops the infinite retry cycle, which hopefully helps resolve the freezing issue you mentioned!
2. **True Error Resumption**: 
   - Introduced `pendingResumeState` to bridge the snapshot to the scheduler.
   - Enhanced `events.go` to handle `m.State != nil` during `EventError`, properly persisting the master list and `.detail.gob`. 
   - Note: Fixed a sparse-snapshot bug where `SaveStateWithOptions` would clobber fields; we now properly backfill the `snapshot` fields before saving.
3. **Enqueue Precheck**: Added a fail-open precheck in `manager.go` before allocating files. It rejects the enqueue if `FileSize + PrecheckSafetyBuffer > FreeSpace`. 
   *(Note: `RuntimeConfig.PrecheckSafetyBuffer` defaults to 500 MiB which works great for our GUI, but you might want to consider adjusting this default for general Surge usage).*

### A quick heads up on the TUI (UX):
The error message string is now properly saved into the master list (`entry.Error = m.Err.Error()`). However, as I mentioned in the issue, I'm not super familiar with the TUI side of things!

Right now, the API (`scheduler.GetStatus` and `local_service.GetStatus`) doesn't bubble up the `DownloadStatus.Error` field yet, and the TUI just displays the standard `StatusError` icon without the text. 
I've left the TUI rendering completely untouched. I figured you'd likely have your own vision for how to display error descriptions cleanly without breaking the current vibe of the UI, so I've just laid the backend groundwork for you to plug into whenever you're ready.

I've included a suite of regression tests covering some edge cases. Take a look when you have a chance, and feel free to adjust or build upon this however you see fit!

Resolves #575

## Summary by CodeRabbit

- **New Features**
  - Added disk-space checks before downloads begin, including a configurable safety buffer.
  - Downloads now recognize disk-full and quota errors across supported platforms.
  - Interrupted downloads preserve more progress and metadata for reliable resumption.

- **Bug Fixes**
  - Prevented retries and fallback behavior for insufficient-disk-space errors.
  - Cancellation and deadline failures now remain distinct and avoid misleading error states.
  - Improved paused and failed download records, including error details, progress, rate limits, and file names.

- **Tests**
  - Expanded coverage for disk-space checks, retry behavior, error persistence, and resume-state handling.

### CODE REVIEW

**Comment by superGekFordJ** [comment]
The bot's findings are super interesting, and its note on `worker.go` (about the missing task range in the snapshot) actually caught me off guard!

It reminded me that in my downstream engine (GoAria), I had built an `abandonedRemaining` stash pool to handle exactly this. When an ENOSPC error occurs, I throw the failed block into this pool before deleting the active state. This way, the snapshot can still collect the downloaded progress without triggering a retry storm by pushing it back to the live queue.

I just checked my GoAria codebase to remember why I didn't include this stash feature in the initial PR. It turns out, to keep the PR as clean and minimal as possible, I intentionally stripped out all of my GoAria-specific worker constructs (like the custom `TieredBufferPool` and this `abandonedRemaining` stash). I assumed Surge's simpler worker model wouldn't need that custom state-saving logic. But as the AI rightly pointed out, without a stash mechanism, the snapshot loses the failing range!

I've been thinking about whether I should try to introduce this stash mechanism to Surge to address this, or if it would be better to handle the bot's finding in a more lightweight way. I'd love to know your architectural preference here.

Also, thanks for the recent refactoring and the Windows `ERROR_HANDLE_DISK_FULL` additions on this PR, those tweaks are awesome! Just a quick note on the bot's other suggestion in `events.go` (using `snapshot.Downloaded < existing.Downloaded`): I feel a simple `max()` comparison might actually be a bit risky if a snapshot is authoritatively lower than the optimistic master list (for example, due to partial chunk rewrites). I'm currently trying to figure out a safer merge policy on my end, so I just wanted to share this thought in case you decide to tackle that bot comment later.

**Comment by superGekFordJ** [comment]
To avoid the risks of a blind max(), I went with a 'task-backed' rule that trusts the snapshot whenever tasks are present, and also made sure to explicitly carry over the URL and DestPath before saving so sparse error or pause events can't accidentally wipe out the resume key."

**Comment by superGekFordJ** [comment]
For reference, here is the patch for our abandonedRemaining stash pool that safely avoids the retry storm issue:
```patch
diff --git a/internal/strategy/concurrent/downloader.go b/internal/strategy/concurrent/downloader.go
index c512c56..489e434 100644
--- a/internal/strategy/concurrent/downloader.go
+++ b/internal/strategy/concurrent/downloader.go
@@ -39,6 +39,11 @@ type ConcurrentDownloader struct {
 	bufPool      sync.Pool
 	Headers      map[string]string // Custom HTTP headers from browser (cookies, auth, etc.)
 	hostLimiter  *transport.HostRateLimiter
+	// abandonedRemaining holds in-flight residuals from workers that failed
+	// with a terminal error (e.g. ENOSPC) and were removed from activeTasks
+	// before saveStateSnapshot could observe them. saveStateSnapshot unions
+	// these with active+queue drains so resume does not lose the failing range.
+	abandonedRemaining []types.Task
 }
 
 // NewConcurrentDownloader creates a new concurrent downloader with all required parameters
@@ -559,7 +564,8 @@ func (d *ConcurrentDownloader) handlePause(destPath string, fileSize int64, queu
 }
 
 func (d *ConcurrentDownloader) saveStateSnapshot(destPath string, fileSize int64, queue *TaskQueue, candidateMirrors []string, emitPauseEvent bool) error {
-	// 1. Collect active tasks as remaining work FIRST
+	// 1. Collect active tasks as remaining work, then drain abandoned residuals
+	// (off-queue stashes from terminal-error workers), then drain the live queue.
 	var activeRemaining []types.Task
 	d.activeMu.Lock()
 	for _, active := range d.activeTasks {
@@ -567,10 +573,12 @@ func (d *ConcurrentDownloader) saveStateSnapshot(destPath string, fileSize int64
 			activeRemaining = append(activeRemaining, *remaining)
 		}
 	}
+	abandoned := d.abandonedRemaining
+	d.abandonedRemaining = nil
 	d.activeMu.Unlock()
 
 	// 2. Collect remaining tasks from queue
-	allTasks := append([]types.Task(nil), activeRemaining...)
+	allTasks := append(append([]types.Task(nil), activeRemaining...), abandoned...)
 	allTasks = append(allTasks, queue.DrainRemaining()...)
 
 	var remainingTasks []types.Task
diff --git a/internal/strategy/concurrent/worker.go b/internal/strategy/concurrent/worker.go
index 357bda5..0e4b4ca 100644
--- a/internal/strategy/concurrent/worker.go
+++ b/internal/strategy/concurrent/worker.go
@@ -115,7 +115,24 @@ func (d *ConcurrentDownloader) worker(ctx context.Context, id int, mirrors []str
 			// Disk full: fail immediately — no in-place retry, no mirror rotation,
 			// no residual requeue. Stricter than permanent HTTP.
 			if types.IsInsufficientDiskSpace(lastErr) {
+				// Stash the in-flight RemainingTask off the live queue before
+				// deleting from activeTasks. saveStateSnapshot unions this stash
+				// with active+queue drains so resume keeps the failing range
+				// without re-Pushing onto the live queue during the storm window.
+				var stash *types.Task
+				if remaining := activeTask.RemainingTask(); remaining != nil {
+					originalEnd := task.Offset + task.Length
+					if remaining.Offset+remaining.Length > originalEnd {
+						remaining.Length = originalEnd - remaining.Offset
+					}
+					if remaining.Length > 0 {
+						stash = remaining
+					}
+				}
 				d.activeMu.Lock()
+				if stash != nil {
+					d.abandonedRemaining = append(d.abandonedRemaining, *stash)
+				}
 				delete(d.activeTasks, id)
 				d.activeMu.Unlock()
 				if d.State != nil {
 ```

**Review by SuperCoolPencil — DISMISSED** [review]
This looks excellent. The `utils/disk_free` logic safely handles missing directory descendants by walking up to existing ancestors, avoids overflow when calculating free bytes, and captures cross-platform OS disk-full errors cleanly (e.g. `syscall.ENOSPC`, `windows.ERROR_DISK_FULL`). Stashing `abandonedRemaining` correctly prevents dropped tasks when workers fail fast due to ENOSPC before `saveStateSnapshot` can read them. All tests pass successfully.

**Comment by superGekFordJ** [comment]
There seems to be a slight misunderstanding—the abandonedRemaining snippet I posted earlier was just a reference and wasn't actually in the PR yet! 

To plug that resume-data-loss hole, I've just pushed 664aa46 to formally implement it. If there's a better approach for the current architecture, we can always revert this and try another way.

File: internal/strategy/concurrent/downloader.go [hunk 1]
```diff
@@ -559,19 +563,22 @@ func (d *ConcurrentDownloader) handlePause(destPath string, fileSize int64, queu
...
 		}
 	}
+	abandoned := d.abandonedRemaining
+	d.abandonedRemaining = nil
 	d.activeMu.Unlock()
 
 	// 2. Collect remaining tasks from queue
-	allTasks := append([]types.Task(nil), activeRemaining...)
-	allTasks = append(allTasks, queue.DrainRemaining()...)
+	allTasks := append(append(append([]types.Task(nil), activeRemaining...), abandoned...), queue.DrainRemaining()...)
```
[the remark this answers is not in the document]
Comment (superGekFordJ) [reply internal/strategy/concurrent/downloader.go:596]: CodeRabbit's concern is a false positive. handlePause and saveStateSnapshot are both called after executeWorkers returns, not concurrently with it.

Each worker has defer wg.Done() (L530) → separate goroutine calls wg.Wait() (L541) → close(workerErrors) (L542) → main goroutine blocks on for err := range workerErrors (L549) → executeWorkers returns (L558) → Download calls handlePause (L311) or saveStateSnapshot (L322).

The cancel() at L534 only accelerates other workers' exit via context cancellation → it does not bypass wg.Wait(). The snapshot is already post-barrier.

**Review by SuperCoolPencil — APPROVED** [review]
Reviewed the final head end to end. Disk-full classification stops worker, fallback, and scheduler retry storms; terminal snapshots retain unfinished ranges and persist resumable error state. Relevant concurrency-heavy packages pass with the race detector, and the full cross-platform CI matrix is green.
